cluster-autoscaler/core/static_autoscaler_test.go

```diff
@@ -2396,6 +2413,240 @@ func TestFilterOutYoungPods(t *testing.T) {
 	}
 }
 
+func TestStaticAutoscalerRunOnceInvokesScaleDownStatusProcessor(t *testing.T) {
+	options := config.AutoscalingOptions{
+		NodeGroupDefaults: config.NodeGroupAutoscalingOptions{
+			ScaleDownUnneededTime:         -1 * time.Nanosecond, // enforce immediate scaledown/drain for ready
+			ScaleDownUnreadyTime:          -1 * time.Nanosecond, // enforce immediate scaledown/drain for unready
+			ScaleDownUtilizationThreshold: 0.5,
+			MaxNodeProvisionTime:          10 * time.Second,
+		},
+		EstimatorName:    estimator.BinpackingEstimatorName,
+		ScaleDownEnabled: true,
+		MaxNodesTotal:    10,
+		MaxCoresTotal:    10,
+		MaxMemoryTotal:   100000,
+	}
+	now := time.Now()
+	n1 := BuildTestNode("n1", 1000, 1000)
+	SetNodeReadyState(n1, true, now)
+	n2 := BuildTestNode("n2", 1000, 1000)
+	SetNodeReadyState(n2, true, now)
+
+	underUtilizedPod := BuildTestPod("p1", 20, 20, WithNodeName("n1"))
+	utilizedPod := BuildTestPod("p1", 800, 800, WithNodeName("n1"))
+
+	testCases := map[string]struct {
+		pods                         []*apiv1.Pod
+		nodes                        []*apiv1.Node
+		fakeDeletionResults          map[string]status.NodeDeleteResult
+		fakeDeletionResultsNodeGroup string
+		expectedStatus               *status.ScaleDownStatus
+	}{
```

Comment:
IMO the test cases test the internals of dependencies more than the RunOnce logic they're supposed to. 

Ideally, for this test we'd like to verify that:
1. The processor always gets called during the loop, regardless of other logic.
2. The processor always gets all relevant data when it's called.

Obviously we can't cover every scenario, but for 1. I could see e.g. 3 cases: "scale-down happened", "only scale-up happened, scale-down didn't because no candidates", "nothing happened, the loop short-circuited early". For 2., we need to verify that in all 3 cases, all relevant pieces of the status are filled correctly. I don't see much value in testing the internals of the status (e.g. different unremovable reasons, or deletion result types), that should be covered by the unit tests of the components where we get the data from. All we need to verify here is that the data was passed from the components correctly (and as much data as it makes sense for every test case).

The test is still good enough IMO, this is a nit/something to consider in the future.

Reply:
I think you are right, it's quite hard to keep concentration on isolation while implementing test which indirectly touches so many parts of CA, thanks for the feedback! 

As a result I've reduced tests count to 3, now it's testing following scenarious: scaledown, no scaledown, no scaledown with deleted node

Reply:
Oh yeah definitely, this line is frequently blurred, and CA is frequently not easy to test because of old tech debt. The test cases are very clear now, thank you!